projects/interfaces/buffer/buffer_test.go

```diff
@@ -0,0 +1,73 @@
+package OurBuffer
+
+import (
+	"io"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+// If you make a buffer named b containing some bytes, calling b.Bytes() returns the same bytes you created it with.
+func TestBufferBytesReturnsInitialBytes(t *testing.T) {
+	inputBytes := []byte("Hello, World!")
+	b := NewBuffer(inputBytes)
+
+	bytes := b.Bytes()
+
+	excepted := string(inputBytes)
+	actual := string(bytes)
+
+	require.Equal(t, excepted, actual)
+}
+
+// If you write some extra bytes to that buffer using b.Write(), a call to b.Bytes() returns both the initial bytes and the extra bytes.
+func TestBufferWriteExtraBytes(t *testing.T) {
+	inputBytes := []byte("Hello ")
+	b := NewBuffer(inputBytes)
+	n, err := b.Write([]byte("World!"))
+	require.NoError(t, err)
+
+	bytes := b.Bytes()
+	expected := string((bytes))
+	actual := string(inputBytes) + "World!"
+	require.Equal(t, expected, actual)
+	require.Equal(t, len(bytes), len(inputBytes)+n)
+}
+
+// If you call b.Read() with a slice big enough to read all of the bytes in the buffer, all of the bytes are read.
+func TestBufferReadFullSlice(t *testing.T) {
+	inputBytes := []byte("Hello, World!")
+	b := NewBuffer(inputBytes)
+
+	outputBytes := make([]byte, len(inputBytes))
+	n, err := b.Read(outputBytes)
+	require.ErrorAs(t, err, &io.EOF)
+
+	expected := string(inputBytes)
+	actual := string(outputBytes)
+	require.Equal(t, expected, actual)
+	require.Equal(t, len(inputBytes), n)
+}
+
+// If you call b.Read() with a slice smaller than the contents of the buffer, some of the bytes are read. If you call it again, the next bytes are read.
+func TestBufferReadPartialSlices(t *testing.T) {
+	inputBytes := []byte("Hello, World!")
+	b := NewBuffer(inputBytes)
+
+	outputBytes := make([]byte, 7)
+	t.Run("first read", func(t *testing.T) {
+		n, err := b.Read(outputBytes)
+		require.NoError(t, err)
+		require.Equal(t, "World!", b.String())
+		require.Equal(t, "Hello, ", string(outputBytes))
+		require.Equal(t, 7, n)
+	})
+
+	t.Run("Second Read", func(t *testing.T) {
+		n, err := b.Read(outputBytes)
+		require.ErrorAs(t, err, &io.EOF)
+		require.Equal(t, b.String(), "")
+		require.Equal(t, "World! ", string(outputBytes))
```

Comment:
This is a really interesting assertion, and I'd be interested to talk about it!

This is reading 6 bytes into a slice which can contain 7 bytes.

You're asserting over all 7 bytes.

I can argue both ways whether we should be asserting over 6 or 7 bytes. Can you talk through why you went with 7? What arguments could you make for asserting over 6? What arguments could you make for 7? How did you decide that the 7th byte should be a space?

Reply:
tbh I was thinking that the space comes from the `copy` function. But just discovered that the space is from the first read. ( from doc of `copy` the source and destination may overlap). 
For asserting over 7 bytes , I think when reading a buffer we work with the same slice whether for the first read or the second. (situation when we loop over the content of the buffer to read it all).
If so , maybe I need to make the `Read` method responsible for adjusting the slice length if `len(p)> len(b.bytes)`. 
If we assert over 6 bytes, we will need to change the slice length manually  when `len(p)> len(b.bytes)` (create a new slice and use it as new parameter for `Read`).

Reply:
I tried this : 
```go
func (b *OurByteBuffer) Read(p []byte) (n int, err error) {
	n = len(p)
	if n > len(b.bytes) {
		n = len(b.bytes)
              // reset the slice
		p = make([]byte, n)
	}

	copy(p, b.bytes[:n])

	b.bytes = b.bytes[n:]

	if len(b.bytes) == 0 {
		err = io.EOF
	}

	return n, err
}
```

But the test is failing because the slice is still holding the first read.

Reply:
> from doc of `copy` the source and destination may overlap

This is actually talking about the source and destination for a single call, i.e. it's saying the following is allowed:
```go
bytes := []byte("hello!")
copy(bytes[1:], bytes)
require.Equal(t,[]byte("hhello"), bytes)
```

You shouldn't need to change the size of the slice, or make a new slice, in your `Read` - in fact, that won't work, because `p` is a copy of a slice (where a slice is a pointer + capacity + length struct-ish thing) so if you make changes to `p` it won't modify the caller's slice (but if you _write to the memory through the pointer_ the contents will get updated, because you're writing to a shared pointer).

But `Read` doesn't promise it will write to the whole slice it's given. In fact, the reason `Read` returns `n int` is so that you as a caller know how much of the passed slice `p` you should actually read. Anything in `p` _after_ `n` will continue to have whatever value it happened to have before `Read` was called.

What do you think this means we should do, both in our `Read` function, and in our test?